File: endpoint-insights-api/src/main/java/com/vsp/endpointinsightsapi/service/NotificationGroupService.java
package com.vsp.endpointinsightsapi.service;

import com.vsp.endpointinsightsapi.model.NotificationGroup;
import com.vsp.endpointinsightsapi.model.entity.NotificationGroupMember;
import com.vsp.endpointinsightsapi.repository.NotificationGroupRepository;
import com.vsp.endpointinsightsapi.repository.NotificationGroupMemberRepository;
import jakarta.transaction.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class NotificationGroupService {

    private static final Logger LOG = LoggerFactory.getLogger(NotificationGroupService.class);

    private final NotificationGroupRepository groupRepository;
    private final NotificationGroupMemberRepository memberRepository;

    public NotificationGroupService(NotificationGroupRepository groupRepository,
                                   NotificationGroupMemberRepository memberRepository) {
        this.groupRepository = groupRepository;
        this.memberRepository = memberRepository;
    }

    // Get all notification groups
    public List<NotificationGroup> getAllGroups() {
        return groupRepository.findAll();
    }

    // Get a specific group by ID
    public Optional<NotificationGroup> getGroupById(UUID id) {
        return groupRepository.findById(id);
    }

    // Create a new notification group
    @Transactional
    public NotificationGroup createGroup(String name, String description, List<String> memberEmails) {
        NotificationGroup group = new NotificationGroup();
        group.setName(name);
        group.setDescription(description);
        group.setMembers(new ArrayList<>());

        NotificationGroup savedGroup = groupRepository.save(group);

        if (memberEmails != null && !memberEmails.isEmpty()) {
            addMembersToGroup(savedGroup.getId(), memberEmails);
        }

        return groupRepository.findById(savedGroup.getId()).orElseThrow();
    }

    // Update an existing group
    @Transactional
    public NotificationGroup updateGroup(UUID id, String name, String description) {
        NotificationGroup group = groupRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Notification group not found: " + id));

        group.setName(name);
        group.setDescription(description);

        return groupRepository.save(group);
    }

    // Delete a group (cascades to delete members)
    @Transactional
    public void deleteGroup(UUID id) {
        if (!groupRepository.existsById(id)) {
            LOG.warn("Attempted to delete non-existent group: {}", id);
            return;
        }
        groupRepository.deleteById(id);
        LOG.info("Deleted notification group: {}", id);
    }

    // Add members to a group
    @Transactional
    public void addMembersToGroup(UUID groupId, List<String> emails) {
        NotificationGroup group = groupRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Notification group not found: " + groupId));

        for (String email : emails) {
            // Check if member already exists
            if (memberRepository.findByGroupIdAndEmail(groupId, email).isEmpty()) {
                NotificationGroupMember member = new NotificationGroupMember();
                member.setGroupId(groupId);
                member.setEmail(email);
                memberRepository.save(member);
            }
        }
    }

    // Remove a member from a group
    @Transactional
    public void removeMemberFromGroup(UUID groupId, String email) {
        memberRepository.findByGroupIdAndEmail(groupId, email)
                .ifPresentOrElse(
                    member -> memberRepository.deleteById(member.getId()),
                    () -> LOG.warn("Member {} not found in group {}", email, groupId)
                );
    }

    // Get all email addresses for a group
    public List<String> getGroupMemberEmails(UUID groupId) {
        List<NotificationGroupMember> members = memberRepository.findAllByGroupId(groupId);
        return members.stream()
                .map(NotificationGroupMember::getEmail)
                .collect(Collectors.toList());
    }

    // Get all member emails for a list of group IDs (used for batch notifications)
    public List<String> resolveGroupsToEmails(List<UUID> groupIds) {
        if (groupIds == null || groupIds.isEmpty()) {
            return new ArrayList<>();
        }

        return groupIds.stream()
                .flatMap(groupId -> getGroupMemberEmails(groupId).stream())
                .distinct()
                .collect(Collectors.toList());
    }
}

File: endpoint-insights-ui/src/app/services/notification-group.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Observable, tap, catchError, throwError } from 'rxjs';
import { environment } from '../../environment';
import { HttpInterceptorService } from './http-interceptor.service';
import { ToastService } from './toast.service';
import { NotificationGroup } from '../models/notification-group.model';

@Injectable({ providedIn: 'root' })
export class NotificationGroupService {
    private readonly httpInterceptService = inject(HttpInterceptorService);
    private readonly toast = inject(ToastService);
    private readonly http = inject(HttpClient);

    // Get all notification groups
    getAllGroups(): Observable<HttpResponse<NotificationGroup[]>> {
        return this.httpInterceptService.get<NotificationGroup[]>(
            `${environment.apiUrl}/notification-groups`
        );
    }

    // Get a specific notification group by ID
    getGroupById(id: string): Observable<NotificationGroup> {
        return this.http.get<NotificationGroup>(
            `${environment.apiUrl}/notification-groups/${id}`
        );
    }

    // Create a new notification group
    createGroup(name: string, description: string, members: string[]): Observable<HttpResponse<NotificationGroup>> {
        const request = {
            name,
            description,
            members: members || []
        };
        return this.httpInterceptService.post<NotificationGroup>(
            `${environment.apiUrl}/notification-groups`,
            request
        ).pipe(
            tap(() => this.toast.onSuccess('Notification group created successfully')),
            catchError(err => {
                this.toast.onError('Failed to create notification group');
                return throwError(() => err);
            })
        );
    }

    // Update an existing notification group
    updateGroup(id: string, name: string, description: string): Observable<HttpResponse<NotificationGroup>> {
        const request = {
            name,
            description
        };
        return this.httpInterceptService.put<NotificationGroup>(
            `${environment.apiUrl}/notification-groups/${id}`,
            request
        ).pipe(
            tap(() => this.toast.onSuccess('Notification group updated successfully')),
            catchError(err => {
                this.toast.onError('Failed to update notification group');
                return throwError(() => err);
            })
        );
    }

    // Delete a notification group
    deleteGroup(id: string): Observable<HttpResponse<void>> {
        return this.httpInterceptService.delete<void>(
            `${environment.apiUrl}/notification-groups/${id}`
        ).pipe(
            tap(() => this.toast.onSuccess('Notification group deleted successfully')),
            catchError(err => {
                this.toast.onError('Failed to delete notification group');
                return throwError(() => err);
            })
        );
    }

    // Add members to a group
    addMembers(groupId: string, emails: string[]): Observable<HttpResponse<void>> {
        const request = {
            emails
        };
        return this.httpInterceptService.post<void>(
            `${environment.apiUrl}/notification-groups/${groupId}/members`,
            request
        ).pipe(
            tap(() => this.toast.onSuccess('Members added to group successfully')),
            catchError(err => {
                this.toast.onError('Failed to add members to group');
                return throwError(() => err);
            })
        );
    }

    // Remove a member from a group
    removeMember(groupId: string, email: string): Observable<HttpResponse<void>> {
        return this.httpInterceptService.delete<void>(
            `${environment.apiUrl}/notification-groups/${groupId}/members/${email}`
        ).pipe(
            tap(() => this.toast.onSuccess('Member removed from group successfully')),
            catchError(err => {
                this.toast.onError('Failed to remove member from group');
                return throwError(() => err);
            })
        );
    }
}
